cpp/include/raft/spatial/knn/detail/ivf_flat/bitonic_sort.cuh

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/cuda_utils.cuh>
+
+namespace raft::spatial::knn::detail::ivf_flat {
+
+namespace helpers {
+
+template <typename T>
+__device__ __forceinline__ void swap(T& x, T& y)
+{
+  T t = x;
+  x   = y;
+  y   = t;
+}
+
+template <typename T>
+__device__ __forceinline__ void assign(bool cond, T* ptr, T x)
+{
+  if (cond) { *ptr = x; }
+}
+
+template <typename T, typename... Ts>
+__device__ __forceinline__ auto first(T x, Ts... xs) -> T
+{
+  return x;
+}
+
+}  // namespace helpers
+
+/**
+ * Bitonic merge at the warp level.
+ *
+ * @tparam Size is the number of elements (must be power of two).
+ * @tparam Ascending is the resulting order (true: ascending, false: descending).
+ */
+template <int Size, bool Ascending>
+struct bitonic_merge {
+  static_assert(isPo2(Size));
+
+  /** How many contiguous elements are processed by one thread. */
+  static constexpr int kArrLen = Size / WarpSize;
+  static constexpr int kStride = kArrLen / 2;
+
+  template <bool Fits, typename Dummy>
+  using when_fits_in_warp =
+    std::enable_if_t<(Fits == (Size <= WarpSize)) && std::is_same_v<Dummy, Dummy>, void>;
```

Comment:
That's used to delay the evaluation of this template till the same time the function templates are being compiled, rather than to compile it together with the enclosing struct.